Abort sorting in case of local agg sort queue overflow

bpintea opened the pull request:
In case the local agg sorter queue gets full and no limit has been provided, the local sorter will now erroneously call the failure callback for every single row in the original rowset that's left over the local queue limit (instead for just the first one).
The failure response is dispatched in any case, so this is relatively harmless.
The sorter continues iterating on the original response fetching subsequent pages. In case of correct Elasticsearch behaviour, this is also harmless, it'll just trigger a number of internal exceptions. However, in case of a pagination defect in Elasticsearch (like #65685, where the same `search_after` is returned), this will result in an effective spin loop, potentially rendering eventually the node unresponsive.

This PR simply breaks both the inner loop iterating over the current unsorted rowset, as well as the outer one, iterating over the left pages.

It also fixes an outdated documentation limitation.

bpintea committed:
Abort sorting in case of local sort overflow

In case the local agg sorter queue gets full, stop iterating on current
result set and fetching next page.

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

matriv reviewed (commented):
Would be great to have some test for that if possible though...

astefan commented:
Is this PR fixing an issue? If so, please provide more details about the reason for this PR.

bpintea commented:
> Is this PR fixing an issue? If so, please provide more details about the reason for this PR.

The PR fixes the behaviour in an error condition, i.e. when the local agg sorter queue gets full and no limit has been provided. In that case the `onFailure` is called, but in a loop, i.e. for every row left in the rowset plus any rows still available in the following page(s). The PR simply breaks the rowset and page loops, it's a minor fix.

I'll try to see if a test is easily implantable, that would make it more obivous.

bpintea committed:
Slight refactor of the fix. Add testing

- Raise an exception on queue insertion failure, instead of running the
  failure callback.
- Add testing for the change.
- Correct stale documentation.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/bail_on_local_sort_overflow

bpintea commented:
I've added testing - a bit stretched maybe, but it tests the fix - and the PR description has been improved.

bpintea committed:
Minor refactoring of testing

- drop Mockito use, not needed in current test form.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/bail_on_local_sort_overflow

bpintea commented:
@elasticmachine run elasticsearch-ci/1 elasticsearch-ci/2 elasticsearch-ci/default-distro elasticsearch-ci/eql-correctness elasticsearch-ci/packaging-sample-windows

bpintea commented:
@elasticmachine run elasticsearch-ci/1 elasticsearch-ci/2 elasticsearch-ci/default-distro elasticsearch-ci/eql-correctness elasticsearch-ci/packaging-sample-windows

matriv commented on docs/reference/sql/limitations.asciidoc:
```diff
@@ -105,8 +105,8 @@ When multiple values are returned for a field, by default, {es-sql} will throw a
-However this type of aggregation does come with a limitation: sorting can only be applied on the key used for the aggregation's buckets. 
-{es-sql} overcomes this limitation by doing client-side sorting however as a safety measure, allows only up to *512* rows.
+However this type of aggregation does come with a limitation: sorting can only be applied on the key used for the aggregation's buckets.
+{es-sql} overcomes this limitation by doing client-side sorting however as a safety measure, allows only up to *65535* rows.
```
We should backport this to reflect the change that happened quite some time ago.

matriv reviewed (approved):
LGTM, Nicely done with the test!

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/execution/search/QuerierTests.java:
```diff
@@ -171,4 +184,85 @@ public void testAggSorting_Randomized() {
         assertEquals(expected.subList(0, queueSize), results);
     }
+
+    public Tuple<Integer, Integer> runLocalAggSorterWithNoLimit(int dataSize) {
```
No need for this one to be public. Also, the method should be moved last in the class.

astefan reviewed (approved):
LGTM. Left one minor comment.

bpintea committed:
Address review comments

- remove "public" method attr and move it to the bottom of class.

bpintea committed:
Merge branch 'fix/bail_on_local_sort_overflow' of github.com:bpintea/elasticsearch into fix/bail_on_local_sort_overflow

bpintea committed:
Merge remote-tracking branch 'upstream/master' into fix/bail_on_local_sort_overflow

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java:
```diff
@@ -225,8 +226,13 @@ public void onResponse(Page page) {
 
-            // 1. consume all pages received
-            consumeRowSet(page.rowSet());
+            try {
```
Since we already have access to the listener, how about returning a bool instead of throwing an exception.
There's nothing wrong with it yet consumeRowSet is used only internally so promoting the condition to an exception only to catch it later  looks a bit wasteful to me.

suggestion:
```
if (consumeRowSet(page.rowSet()) == false) {
   return;
}
```

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/execution/search/QuerierTests.java:
```diff
@@ -171,4 +184,85 @@ public void testAggSorting_Randomized() {
+
+            @Override
+            public Schema schema() {
+                return new Schema(Collections.<String>emptyList(), Collections.<DataType>emptyList());
```
Static imports?

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/execution/search/QuerierTests.java:
```diff
@@ -171,4 +184,85 @@ public void testAggSorting_Randomized() {
+        assertEquals("Exactly one failure expected", 1, actions.v2().intValue());
+    }
+
+    Tuple<Integer, Integer> runLocalAggSorterWithNoLimit(int dataSize) {
```
Nice

costin reviewed (approved):
LGTM

bpintea committed:
Address review comments

- revert back to bool-based flow control.
- import static methods where possible.

bpintea committed:
Style fix

- unused import removed.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/bail_on_local_sort_overflow
